tests/tests.py

```diff
@@ -938,3 +952,425 @@ def test_invalidate_model_with_geometry(self):
+    pass
+
+
+class TransactionalInvalidationTests(BaseTestCase):
+    fixtures = ['basic']
+
+    def test_atomic_block_change_and_invalidate_obj(self):
+        with self.assertNumQueries(1):
+            self.assertEqual('Django', list(Category.objects.filter(pk=1).cache())[0].title)
+        with self.assertNumQueries(0):
+            self.assertEqual('Django', list(Category.objects.filter(pk=1).cache())[0].title)
+        with atomic():
+            self.assertTrue(in_transaction())
+            with self.assertNumQueries(1):
+                self.assertEqual('Django', list(Category.objects.filter(pk=1).cache())[0].title)
+            self.assertEqual(
+                'Django',
+                return_from_other_thread(
+                    lambda: list(Category.objects.filter(pk=1).cache())[0].title
+                )
+            )
+            obj = list(Category.objects.filter(pk=1).cache())[0]
+            obj.title = 'Changed'
+            obj.save()
+            invalidate_obj(obj)
+            with self.assertNumQueries(1):
+                self.assertEqual('Changed', list(Category.objects.filter(pk=1).cache())[0].title)
+            self.assertEqual(
+                'Django',
+                return_from_other_thread(
+                    lambda: list(Category.objects.filter(pk=1).cache())[0].title
+                )
+            )
+            queue_item = list(_function_queue)[-1]
+            self.assertEqual('invalidate_dict', queue_item['func'].__name__)
+            self.assertEqual(obj.__class__, queue_item['args'][0])
+        self.assertFalse(in_transaction())
+        self.assertEqual(
+            'Changed',
+            return_from_other_thread(
+                target=lambda: list(Category.objects.filter(pk=1).cache())[0].title,
+                queries=1
+            )
+        )
+        with self.assertNumQueries(0):
+            self.assertEqual('Changed', list(Category.objects.filter(pk=1).cache())[0].title)
+
+    def test_nested_atomic_block_change_and_invalidate_model(self):
+        with self.assertNumQueries(1):
+            self.assertEqual('Django', list(Category.objects.filter(pk=1).cache())[0].title)
+        with self.assertNumQueries(0):
+            self.assertEqual('Django', list(Category.objects.filter(pk=1).cache())[0].title)
+        with atomic():
+            with atomic():
+                self.assertTrue(in_transaction())
+                with self.assertNumQueries(1):
+                    self.assertEqual('Django', list(Category.objects.filter(pk=1).cache())[0].title)
+                self.assertEqual(
+                    'Django',
+                    return_from_other_thread(
+                        lambda: list(Category.objects.filter(pk=1).cache())[0].title
+                    )
+                )
+                obj = list(Category.objects.filter(pk=1).cache())[0]
+                obj.title = 'Changed'
+                obj.save()
+                invalidate_model(obj.__class__)
```

Comment:
If you really want to test `invalidate_model()` then you should use non-invalidating update:
```python
Category.objects.filter(pk=1).update(title='Change')
```
But I think this test is unnecessary.